src/ResourceManagement/ResourceManager/GenericResource/GenericResourcesImpl.cs

```diff
@@ -158,13 +183,20 @@ public new IGenericResource GetById(string id)
 
         public async new Task<IGenericResource> GetByIdAsync(string id, CancellationToken cancellationToken = default(CancellationToken))
         {
+            string apiVersion = ResourceUtils.DefaultApiVersionFromResourceId(
+                id, 
+                await Manager.Providers.GetByNameAsync(ResourceUtils.ResourceProviderFromResourceId(id), cancellationToken));
```

Comment:
Am I right in thinking that `IProviders.GetByNameAsync()` makes a web request? If so, my one concern about this approach is that it's going to change this method from making a single request for the resource to two requests:

1. To get the `IProvider` instance.
2. To actually get the resource using the ApiVersion we've just looked up.

Would it be worth introducing some kind of caching to make sure that you don't have to lookup the version / the `IProvider` every single time? Otherwise I think it might end up adding quite a lot of overhead to this request, as well as using up more of the user's API rate limit.

Reply:
Added cache for API versions.